File: 2022/solver.py
#!/usr/bin/env python3

import os
import sys
import importlib

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print(f"Usage: {os.path.basename(__file__)} day input_type [part]")
        sys.exit(1)

    day = sys.argv[1]
    input_type = sys.argv[2]
    part = sys.argv[3] if len(sys.argv) > 3 else None

    formatted_day = day.zfill(2)

    module = importlib.import_module(f"{formatted_day}.solutions", ".")

    dir = os.path.dirname(os.path.realpath(__file__))
    with open(f"{dir}/{formatted_day}/{input_type}.txt", "r") as f:
        data = f.read()

    if part == "1" or part is None:
        print(module.part1(data))
    if part == "2" or part is None:
        print(module.part2(data))

File: 2023/solver.rb
# frozen_string_literal: true

Dir[File.join(__dir__, 'day??', 'solutions.rb')].each { |file| require file }

if ARGV.length < 2
  puts "Usage: #{__FILE__} day input_type [part]"
  exit(1)
end

day = format('%02d', ARGV[0])
input_type = ARGV[1]
part = ARGV[2].to_i

puts "Running day #{day.to_i}..."

input = File.read(File.join(__dir__, "day#{day}", "#{input_type}.txt"))

if [0, 1].include?(part)
  # Dynamic version of: Day01.part1(input)
  result = Object.const_get("Day#{day}").public_send(:part1, input)
  puts "Part 1: #{result}"
end

if [0, 2].include?(part)
  result = Object.const_get("Day#{day}").public_send(:part2, input)
  puts "Part 2: #{result}"
end
